File: lib/models/user_model.dart
class User {
  final String? id;
  final String nome;
  final String email;
  final String matricula;
  final String role;
  final String? curso;
  final List<String> turmas;
  final int saldo;
  final String? fotoPerfil;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  User({
    this.id,
    required this.nome,
    required this.email,
    required this.matricula,
    required this.role,
    this.curso,
    required this.turmas,
    required this.saldo,
    this.fotoPerfil,
    this.createdAt,
    this.updatedAt,
  });

  factory User.fromJson(Map<String, dynamic> json) {
    return User(
      id: json['_id'],
      nome: json['nome'],
      email: json['email'],
      matricula: json['matricula'],
      role: json['role'],
      curso: json['curso'],
      turmas: List<String>.from(json['turmas'] ?? []),
      saldo: json['saldo'] ?? 0,
      fotoPerfil: json['fotoPerfil'],
      createdAt: json['createdAt'] != null 
          ? DateTime.parse(json['createdAt']) 
          : null,
      updatedAt: json['updatedAt'] != null 
          ? DateTime.parse(json['updatedAt']) 
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'nome': nome,
      'email': email,
      'matricula': matricula,
      'role': role,
      'curso': curso,
      'turmas': turmas,
      'saldo': saldo,
      'fotoPerfil': fotoPerfil,
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }

  User copyWith({
    String? id,
    String? nome,
    String? email,
    String? matricula,
    String? role,
    String? curso,
    List<String>? turmas,
    int? saldo,
    String? fotoPerfil,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return User(
      id: id ?? this.id,
      nome: nome ?? this.nome,
      email: email ?? this.email,
      matricula: matricula ?? this.matricula,
      role: role ?? this.role,
      curso: curso ?? this.curso,
      turmas: turmas ?? this.turmas,
      saldo: saldo ?? this.saldo,
      fotoPerfil: fotoPerfil ?? this.fotoPerfil,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

class LoginRequest {
  final String matricula;
  final String senha;

  LoginRequest({
    required this.matricula,
    required this.senha,
  });

  Map<String, dynamic> toJson() {
    return {
      'matricula': matricula,
      'senha': senha,
    };
  }
}

class LoginResponse {
  final String token;
  final User user;

  LoginResponse({
    required this.token,
    required this.user,
  });

  factory LoginResponse.fromJson(Map<String, dynamic> json) {
    return LoginResponse(
      token: json['token'],
      user: User.fromJson(json['user']),
    );
  }
} 

File: backend/routes/userModel.js
const mongoose = require('mongoose');
const bcrypt = require('bcryptjs');

const userSchema = new mongoose.Schema({
    nome: {
        type: String,
        required: [true, 'Nome é obrigatório'],
        trim: true
    },
    email: { 
        type: String, 
        unique: true,
        required: [true, 'Email é obrigatório'],
        lowercase: true,
        trim: true,
        match: [/^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/, 'Email inválido']
    },
    senha: {
        type: String,
        required: [true, 'Senha é obrigatória'],
        minlength: [6, 'Senha deve ter pelo menos 6 caracteres']
    },
    matricula: { 
        type: String, 
        unique: true,
        required: [true, 'Matrícula é obrigatória'],
        trim: true
    },
    role: { 
        type: String, 
        enum: ['aluno', 'professor', 'admin'], 
        default: 'aluno' 
    },
    curso: { 
        type: String, 
        enum: ['Engenharia de Alimentos', 'Agropecuária', 'Informática para Internet'],
        required: function() {
            return this.role === 'aluno';
        }
    },
    turmas: [String],
    saldo: { 
        type: Number, 
        default: 0,
        min: [0, 'Saldo não pode ser negativo']
    },
    fotoPerfil: { 
        type: String, 
        default: '' 
    },
    ultimoLogin: {
        type: Date,
        default: Date.now
    },
    ativo: {
        type: Boolean,
        default: true
    }
}, 
{ timestamps: true });

// Middleware para hash da senha antes de salvar
userSchema.pre('save', async function(next) {
    // Só hash a senha se ela foi modificada (ou é nova)
    if (!this.isModified('senha')) return next();
    
    try {
        // Hash da senha com salt de 12 rounds
        const salt = await bcrypt.genSalt(12);
        this.senha = await bcrypt.hash(this.senha, salt);
        next();
    } catch (error) {
        next(error);
    }
});

// Método para comparar senhas
userSchema.methods.compararSenha = async function(senhaCandidata) {
    return await bcrypt.compare(senhaCandidata, this.senha);
};

// Método para adicionar coins
userSchema.methods.adicionarCoins = function(quantidade) {
    if (quantidade > 0) {
        this.saldo += quantidade;
        return this.save();
    }
    throw new Error('Quantidade deve ser positiva');
};

// Método para remover coins
userSchema.methods.removerCoins = function(quantidade) {
    if (quantidade > 0 && this.saldo >= quantidade) {
        this.saldo -= quantidade;
        return this.save();
    }
    throw new Error('Saldo insuficiente ou quantidade inválida');
};

// Método para atualizar último login
userSchema.methods.atualizarUltimoLogin = function() {
    this.ultimoLogin = new Date();
    return this.save();
};

// Método para retornar dados públicos (sem senha)
userSchema.methods.toPublicJSON = function() {
    const userObject = this.toObject();
    delete userObject.senha;
    return userObject;
};

// Índices para melhor performance
userSchema.index({ email: 1 });
userSchema.index({ matricula: 1 });
userSchema.index({ role: 1 });
userSchema.index({ ativo: 1 });

module.exports = mongoose.model('User', userSchema);
